src/System.Private.ServiceModel/tools/test/SelfHostWcfService/WSDuplexService.cs

```diff
@@ -0,0 +1,198 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.ServiceModel;
+using System.Threading;
+
+namespace WcfService
+{
+    [ServiceBehavior(InstanceContextMode = InstanceContextMode.Single, ConcurrencyMode = ConcurrencyMode.Reentrant)]
+    public class WSDuplexService : IWSDuplexService
+    {
+        private string ContentToReplace = "ContentToReplace";
+        private string ResponseReplaceThisContent = "ResponseReplaceThisContent";
+        private string ReplacedContent = "ReplacedContent";
+        private string LastMessage = "LastMessage";
+
+        private string exceptionstring = string.Empty;
+        List<string> log = new List<string>();
+        static bool continuePushingData;
+
+        static int seed = DateTime.Now.Millisecond;
+        static Random rand = new Random(seed);
+
+        static FlowControlledStream localStream;
+
+        public void UploadData(string data)
+        {
+            if (data.Contains(ContentToReplace) || data.Contains(ReplacedContent) || data.Contains(ResponseReplaceThisContent))
+            {
+                log.Add(string.Format("UploadData received {0}", data));
+            }
```

Comment:
Note for the future only -- I believe @shhsu was considering adding a logging layer.  When that happens, it might make sense to use that mechanism.  I also see an interesting issue here -- this service is single instance and reentrant.  I believe that implies it needs to be thread-safe when it touches the log List.  If multiple tests every hit this same endpoint at the same time, we risk fatal exceptions on the server side.